### PR TITLE
Dynamo paging

### LINKED ISSUE #12: DynamoDB scan operation ExclusiveStartKey
The default scan method assumes that the returned results do not exceed 1MB. However, if the result returned by the scan operation exceeds 1MB this method requires keeping track of the [ExclusiveStartKey](http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html#API_Scan_RequestSyntax) and then iterating over the result based on that. please refer to the following code snippet for example:
```scala
def scan2(table: Table,
    filter: Seq[(String, aws.model.Condition)],
    limit: Int = 1000,
    segment: Int = 0,
    totalSegments: Int = 1,
    select: Select = aws.model.Select.ALL_ATTRIBUTES,
    attributesToGet: Seq[String] = Nil): Seq[Item] = try {

    val req = new aws.model.ScanRequest()
      .withTableName(table.name)
      .withScanFilter(filter.toMap.asJava)
      .withSelect(select)
      .withLimit(limit)
      .withSegment(segment)
      .withTotalSegments(totalSegments)
    if (!attributesToGet.isEmpty) {
      req.setAttributesToGet(attributesToGet.asJava)
    }

    
     implicit val dynamoDB = DynamoDB(credentialsawsscala)
     val scanResult = dynamoDB.scan(req)
     
     
     var lastEvaluatedKey = scanResult.getLastEvaluatedKey()
     
     var all_data = scala.collection.mutable.Seq[Item]()
     
     val single_iteration = scanResult.getItems.asScala.map(i => Item(table, i)).toSeq
    
     all_data = all_data ++ single_iteration
     while (lastEvaluatedKey!=null)
     {
    
      val req: com.amazonaws.services.dynamodbv2.model.ScanRequest = new aws.model.ScanRequest()
      .withTableName(table.name)
      .withScanFilter(filter.toMap.asJava)
      .withSelect(select)
      .withLimit(limit)
      .withSegment(segment)
      .withTotalSegments(totalSegments)
      .withExclusiveStartKey(lastEvaluatedKey)
      
      val scanResult = dynamoDB.scan(req)
      val single_iteration = scanResult.getItems.asScala.map(i => Item(table, i)).toSeq
      all_data = all_data ++ single_iteration
      lastEvaluatedKey = scanResult.getLastEvaluatedKey()
     }
     all_data 
    
  } catch { case e: aws.model.ResourceNotFoundException => Nil }
```

Please let me know how do you want to address this issue, I will be happy to send a pull request .

### PR DESCRIPTION
**Pull request by marhel**
This PR adds paging support to the DynamoDB wrapper. As per [the docs](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/QueryAndScan.html#Pagination) one must check if the returned QueryResult/ScanResult has a non-null lastEvaluatedKey, in which case Dynamo has only given you partial results for your scan/query, and you must then reissue the query with a corresponding ExclusiveStartKey.

Scans become partial (and paging is needed) if Dynamo needs to scans more than the user-specified 'limit' items, or scans more than 1MB of data. This means that scans over anything but very small tables almost always becomes paged. Queries operate over much less data due to hashPK and rangePK, but with many matching items, will also need paging.

Closes #12

### CODE REVIEW

File: src/main/scala/awscala/dynamodbv2/DynamoDB.scala [hunk 1]
```diff
@@ -340,13 +352,104 @@ trait DynamoDB extends aws.AmazonDynamoDB {
...
       req.setAttributesToGet(attributesToGet.asJava)
     }
 
-    scan(req).getItems.asScala.map(i => Item(table, i)).toSeq
+    val pager = new ScanResultPager(table, scan(_), req, pageStatsCallback)
+    pager.toSeq // will return a Stream[Item]
   } catch { case e: aws.model.ResourceNotFoundException => Nil }
+}
+
+case class PageStats(page: Int, lastPage: Boolean, limit: Int, scanned: Int, items: Int, consumedCapacity: aws.model.ConsumedCapacity)
```
Comment (seratch) [inline src/main/scala/awscala/dynamodbv2/DynamoDB.scala]: Please separate source file except sealed trait and its sub types. (e.g.) move this case class to `PageStats.scala`.

**Comment by seratch** [comment]
Thank you for your contribution. The feature looks good to me. After doing the following things, I'm willing to merge this.

- separate source code for each class/object
- squash commits into one

**Comment by marhel** [comment]
OK, so I've split out PageStats into PageStats.scala and ResultPager, QueryResultPager and ScanResultPager into ResultPager.scala. Or did you want QueryResultPager and ScanResultPager into separate files as well? I also made ResultPager a sealed trait.

Regarding squashing, if you want PRs squashed, since this March you can squash the PR yourself upon merging, see https://help.github.com/articles/about-pull-request-merge-squashing/.

But please let me know if you want me to squash the PR anyway!

**Comment by marhel** [comment]
Ping

**Comment by seratch** [comment]
Oops, thanks.

**Pull request merged**